pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -815,30 +968,42 @@ func TestReconcileOnCompletedPipelineRun(t *testing.T) {
 	defer cancel()
 	c := testAssets.Controller
 	clients := testAssets.Clients
+	reconciler := c.Reconciler.(*Reconciler)
+	fr := reconciler.Recorder.(*record.FakeRecorder)
 
-	if err := c.Reconciler.Reconcile(context.Background(), "foo/test-pipeline-run-completed"); err != nil {
+	if err := reconciler.Reconcile(context.Background(), "foo/test-pipeline-run-completed"); err != nil {
 		t.Fatalf("Error reconciling: %s", err)
 	}
 
-	if len(clients.Pipeline.Actions()) != 2 {
+	actions := clients.Pipeline.Actions()
+	if len(actions) != 2 {
+		t.Errorf("# Actions: %d, Actions: %#v", len(actions), actions)
 		t.Fatalf("Expected client to have updated the TaskRun status for a completed PipelineRun, but it did not")
 	}
 
-	_, ok := clients.Pipeline.Actions()[1].(ktesting.UpdateAction).GetObject().(*v1beta1.PipelineRun)
-	if !ok {
-		t.Errorf("Expected a PipelineRun to be updated, but it wasn't.")
-	}
-	t.Log(clients.Pipeline.Actions())
-	actions := clients.Pipeline.Actions()
+	pipelineUpdates := 0
 	for _, action := range actions {
 		if action != nil {
-			resource := action.GetResource().Resource
-			if resource == "taskruns" {
-				t.Fatalf("Expected client to not have created a TaskRun for the completed PipelineRun, but it did")
+			switch {
+			case action.Matches("create", "taskruns"):
+				t.Errorf("Expected client to not have created a TaskRun, but it did")
+			case action.Matches("update", "pipelineruns"):
+				pipelineUpdates++
+			case action.Matches("get", "pipeline"):
+				// We expect one pipelinerun update and one pipeline get.
+				// We need to check the pipeline spec in case there are results to be calculated
+				fallthrough
+			default:
+				continue
```

Comment:
this seems pretty complicated for a unit test - should we have unit tests for this as well?

Reply:
Perhaps we could have an helper to check actions and use it in all tests... and we could unit test the helper.

Reply:
But I would do that in a different PR

Reply:
https://github.com/tektoncd/pipeline/issues/2772